File: articles/evolution-history-of-life/c/evolution_kernel.c
/*
 * Compact evolution kernel in C.
 */

#include <math.h>
#include <stdio.h>

double mean_fitness(double p, double w_AA, double w_Aa, double w_aa) {
    double q = 1.0 - p;
    return p * p * w_AA + 2.0 * p * q * w_Aa + q * q * w_aa;
}

double selection_update(double p, double w_AA, double w_Aa, double w_aa) {
    double q = 1.0 - p;
    double wbar = mean_fitness(p, w_AA, w_Aa, w_aa);
    return (p * p * w_AA + p * q * w_Aa) / wbar;
}

double jukes_cantor(double d) {
    if (d >= 0.75) {
        return NAN;
    }

    return -(3.0 / 4.0) * log(1.0 - (4.0 / 3.0) * d);
}

int main(void) {
    double p = 0.2;
    double p_next = selection_update(p, 1.12, 1.05, 1.0);
    double lambda = 0.10;
    double mu = 0.03;
    double r = lambda - mu;

    printf("mean_fitness=%.6f\n", mean_fitness(p, 1.12, 1.05, 1.0));
    printf("p_next=%.6f\n", p_next);
    printf("delta_p=%.6f\n", p_next - p);
    printf("jukes_cantor=%.6f\n", jukes_cantor(0.15));
    printf("net_diversification=%.6f\n", r);
    printf("expected_richness=%.6f\n", 8.0 * exp(r * 50.0));

    return 0;
}

File: articles/evolution-history-of-life/fortran/evolution_kernel.f90
! Compact evolution and deep-time numerical kernel in Fortran.

program evolution_kernel
  implicit none

  real :: p
  real :: q
  real :: w_AA
  real :: w_Aa
  real :: w_aa
  real :: wbar
  real :: p_next
  real :: d
  real :: jc
  real :: lambda
  real :: mu
  real :: r
  real :: n0
  real :: t
  real :: nt

  p = 0.2
  q = 1.0 - p

  w_AA = 1.12
  w_Aa = 1.05
  w_aa = 1.0

  wbar = p ** 2 * w_AA + 2.0 * p * q * w_Aa + q ** 2 * w_aa
  p_next = (p ** 2 * w_AA + p * q * w_Aa) / wbar

  print *, "Mean fitness:", wbar
  print *, "p_next:", p_next
  print *, "delta_p:", p_next - p

  d = 0.15
  jc = -(3.0 / 4.0) * log(1.0 - (4.0 / 3.0) * d)

  print *, "Jukes-Cantor distance:", jc

  lambda = 0.10
  mu = 0.03
  r = lambda - mu
  n0 = 8.0
  t = 50.0
  nt = n0 * exp(r * t)

  print *, "Net diversification:", r
  print *, "Expected richness:", nt
end program evolution_kernel
